### PR TITLE
minor fixes for 4.1.0a1

### PR DESCRIPTION
**Pull request by CarlGao4**
I still have a suggestion: trying all decoders of torchaudio before falling back to ffmpeg (torchaudio 2.1.0 will support ffmpeg backend itself so maybe reading with ffmpeg can be removed?)

image

<details>
<summary>Changes</summary>

> print out the exception when calling callback
> ensures all threads can be stopped when interrupting separation
> add release data for 4.0.1

</datails>

### CODE REVIEW

File: demucs/apply.py [hunk 1]
```diff
@@ -323,17 +319,17 @@ def apply_model(model: tp.Union[BagOfModels, Model],
...
             out = model(padded_mix)
         with lock:
             try:
                 callback(_replace_dict(callback_arg, ("state", "end")))  # type: ignore
             except KeyboardInterrupt:
-                raise
+                return None
             except Exception:
-                pass
+                traceback.print_exc(file=sys.stderr)
```
Comment (adefossez) [inline demucs/apply.py:333]: I think you shouldn't catch the exception and proceed silently, nor print the traceback, you should just let the exception be raised, e.g. just remove the `except Exception` bit.

**Review by adefossez — CHANGES_REQUESTED** [review]

File: demucs/api.py [hunk 2]
```diff
@@ -264,6 +264,26 @@ def separate_tensor(
         """
         if sr is not None and sr != self.samplerate:
             wav = convert_audio(wav, sr, self._samplerate, self._audio_channels)
+        if wav.max() == wav.min():
```
Comment (adefossez) [inline demucs/api.py:267]: I don't get why this is necessary ?

File: demucs/api.py [hunk 3]
```diff
@@ -264,6 +264,26 @@ def separate_tensor(
...
+                        ("shift_idx", self._shifts - 1),
+                        ("segment_offset", wav.shape[1]),
+                        ("audio_length", wav.shape[1]),
+                        ("models", len(self._model.models) if isinstance(self._model, BagOfModels)
+                         else 1)
+                    )
+                )
+            except KeyboardInterrupt:
+                raise
+            except Exception:
```
Comment (adefossez) [inline demucs/api.py:284]: no silent catching of exception, this should be removed. Then both except :  block should be removed.

File: demucs/apply.py [hunk 4]
```diff
@@ -321,19 +315,21 @@ def apply_model(model: tp.Union[BagOfModels, Model],
...
             try:
-                callback(_replace_dict(callback_arg, ("state", "end")))  # type: ignore
+                if callback is not None:
+                    callback(_replace_dict(callback_arg, ("state", "end")))  # type: ignore
             except KeyboardInterrupt:
-                raise
+                return None
             except Exception:
-                pass
+                raise
```
Comment (adefossez) [inline demucs/apply.py:333]: except followed by raise is a noop, just remove line 332 and 333.

File: demucs/apply.py [hunk 5]
```diff
@@ -321,19 +315,21 @@ def apply_model(model: tp.Union[BagOfModels, Model],
...
             try:
-                callback(_replace_dict(callback_arg, ("state", "start")))  # type: ignore
+                if callback is not None:
+                    callback(_replace_dict(callback_arg, ("state", "start")))  # type: ignore
             except KeyboardInterrupt:
-                raise
+                return None
             except Exception:
-                pass
+                raise
```
Comment (adefossez) [inline demucs/apply.py:323]: same as below

File: demucs/api.py (hunk 2 above)
Comment (CarlGao4) [reply demucs/api.py:267]: If `wav.max() == wav.min()`, then `ref.std()` would be 0, which will cause L289 raise `ZeroDivisionError`

File: docs/api.md [hunk 6]
```diff
@@ -127,7 +127,7 @@ progress: If true, show a progress bar.
 
 ##### Notes for callback
 
-The function will be called with only one positional parameter whose type is `dict`. The `callback_arg` will be combined with information of current separation progress. The progress information will override the values in `callback_arg` if same key has been used. To abort the separation, raise `KeyboardInterrupt`.
+The function will be called with only one positional parameter whose type is `dict`. The `callback_arg` will be combined with information of current separation progress. The progress information will override the values in `callback_arg` if same key has been used. To abort the separation, raise `KeyboardInterrupt`. This will directly make the function return `None` but not stop the program.
```
Comment (adefossez) [inline docs/api.md:130]: revert change to doc as mentioned in the other message.

**Review by adefossez — CHANGES_REQUESTED** [review]

File: demucs/apply.py [hunk 7]
```diff
@@ -321,19 +315,17 @@ def apply_model(model: tp.Union[BagOfModels, Model],
...
         with lock:
             try:
-                callback(_replace_dict(callback_arg, ("state", "start")))  # type: ignore
+                if callback is not None:
+                    callback(_replace_dict(callback_arg, ("state", "start")))  # type: ignore
             except KeyboardInterrupt:
-                raise
-            except Exception:
-                pass
+                return None
```
[the remark this answers is not in the document]
Comment (CarlGao4) [reply demucs/apply.py:321]: I used this also to ensure that the pool can be stopped. When receiving normal Ctrl+C, multiprocessing applications can't stop properly.

File: demucs/api.py (hunk 2 above)
Comment (CarlGao4) [reply demucs/api.py:267]: Doing this will make it much slower
It seems no use separating an empty audio

**Review by adefossez — APPROVED** [review]

**Comment by gillbates** [comment]
it seemed 4.1.0a1 is not released in pip?

**Comment by CarlGao4** [comment]
Beta versions will not be pushed to pip to avoid instability. You can install it using git repo.

**Pull request merged**